File: 백준/src/baekjoon/Problem12865.java
package baekjoon;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.Comparator;
import java.util.StringTokenizer;

/**
 * 출처: https://www.acmicpc.net/problem/12865
 */
public class Problem12865 {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		int N = Integer.parseInt(st.nextToken());
		int K = Integer.parseInt(st.nextToken());

		int[][] arr = new int[N + 1][2];
		int[][] dp = new int[N + 1][K + 1];
		for (int i = 1; i <= N; i++) {
			st = new StringTokenizer(br.readLine());
			arr[i][0] = Integer.parseInt(st.nextToken()); // W
			arr[i][1] = Integer.parseInt(st.nextToken()); // K
		}

		Arrays.sort(arr, Comparator.comparingInt(a -> a[0]));

		for (int j = 1; j <= K; j++) {
			for (int i = 1; i <= N; i++) {
				dp[i][j] = Math.max(dp[i][j - 1], dp[i - 1][j]);
				if (j - arr[i][0] >= 0) {
					dp[i][j] = Math.max(dp[i][j], dp[i - 1][j - arr[i][0]] + arr[i][1]);
				}
			}
		}
		System.out.println(dp[N][K]);
	}
}

File: 백준/src/baekjoon/Problem12865.py
import sys

input = sys.stdin.readline

N, K = map(int, input().split())
items = [list(map(int, input().split())) for _ in range(N)]
items.append([0, 0])
items.sort()

dp = [[0] * (N + 1) for _ in range(K + 1)]

for c in range(1, N + 1):
    w, v = items[c]
    for r in range(1, K + 1):
        if w <= r:
            dp[r][c] = max(dp[r][c - 1], dp[r - w][c - 1] + v)
        else:
            dp[r][c] = dp[r][c - 1]

print(dp[K][N])

